components/modules/matrix.c

```diff
@@ -0,0 +1,528 @@
+    }
+  }
+
+  return result;
+}
+
+static void callback_callOne(lua_State* L, int cb, int mask, int arg, uint32_t time)
+{
+  if (cb != LUA_NOREF) {
+    lua_rawgeti(L, LUA_REGISTRYINDEX, cb);
+
+    lua_pushinteger(L, mask);
+    lua_pushvalue(L, arg - 2);
+    lua_pushinteger(L, time);
+
+    luaL_pcallx(L, 3, 0);
+  }
+}
+
+static void callback_call(lua_State* L, DATA *d, int cbnum, int key, uint32_t time)
+{
+  if (d) {
+    lua_rawgeti(L, LUA_REGISTRYINDEX, d->character_ref);
+    lua_rawgeti(L, -1, key);
+    if (lua_type(L, -1) != LUA_TNIL) {
+      callback_callOne(L, d->callback[cbnum], 1 << cbnum, -1, time);
+    }
+    lua_pop(L, 2);
+  }
+}
+
+static void getpins(lua_State *L, int argno, int count, uint8_t *dest)
+{
+  for (int i = 1; i <= count; i++) {
+    lua_rawgeti(L, argno, i);
+    *dest++ = lua_tonumber(L, -1);
+    lua_pop(L, 1);
+  }
+}
+
+// Lua: setup({cols}, {rows}, {characters})
+static int lmatrix_setup( lua_State* L )
+{
+  luaL_checktype(L, 1, LUA_TTABLE);
+  luaL_checktype(L, 2, LUA_TTABLE);
+  luaL_checktype(L, 3, LUA_TTABLE);
+
+  // Get the sizes of the first two tables
+  size_t columns = lua_rawlen(L, 1);
+  size_t rows = lua_rawlen(L, 2);
+
+  if (columns > 255 || rows > 255 || !rows || !columns) {
+    return luaL_error(L, "Number of rows or columns out of range");
+  }
+
+  DATA *d = (DATA *)lua_newuserdata(L, sizeof(DATA) + rows + columns);
+  if (!d) return luaL_error(L, "not enough memory");
+  memset(d, 0, sizeof(*d) + rows + columns);
+  luaL_getmetatable(L, "matrix.keyboard");
+  lua_setmetatable(L, -2);
+
+  // create a self_ref to prevent GC
+
+  lua_pushvalue(L, -1);
+  d->self_ref = luaL_ref(L, LUA_REGISTRYINDEX);
+
+  d->columns = (uint8_t *) (d + 1);
+  d->rows = d->columns + columns;
+  d->column_count = columns;
+  d->row_count = rows;
+
+  esp_timer_create_args_t timer_args = {
+    .callback = lmatrix_timer_done,
+    .dispatch_method = ESP_TIMER_TASK,
+    .name = "matrix_timer",
+    .arg = d
+  };
+
+  d->open = true;
+
+  esp_timer_create(&timer_args, &d->timer_handle);
+
+  for (int i = 0; i < CALLBACK_COUNT; i++) {
+    d->callback[i] = LUA_NOREF;
+  }
+  getpins(L, 1, columns, d->columns);
+  getpins(L, 2, rows, d->rows);
+  lua_pushvalue(L, 3);
+  d->character_ref = luaL_ref(L, LUA_REGISTRYINDEX);
+
+  for (int i = 0; i < d->row_count; i++) {
+    gpio_isr_handler_add(d->rows[i], matrix_interrupt, d);
+  }
+  initialize_pins(L, d);
+
+  return 1;
+}
+
+// Lua: close( )
+static int lmatrix_close( lua_State* L )
+{
+  DATA *d = (DATA *)luaL_checkudata(L, 1, "matrix.keyboard");
+
+  if (d->open) {
+    callback_free(L, d, MATRIX_ALL);
+
+    if (matrix_close( d )) {
+      return luaL_error( L, "Unable to close switch." );
+    }
+
+    esp_timer_stop(d->timer_handle);
+    esp_timer_delete(d->timer_handle);
+    luaL_unref(L, LUA_REGISTRYINDEX, d->character_ref);
+
+    if (!HAS_QUEUED_DATA(d)) {
+      luaL_unref(L, LUA_REGISTRYINDEX, d->self_ref);
+    }
```

Comment:
I'd be tempted to change this to a task_post so we only have a single place where we release the self_ref, for easier reasoning.